File: maximum-subarray.cpp
class Solution {
public:
    int cross(int l,int m,int r,vector<int>& nums)
    {
        int left_sum = INT_MIN,sum=0;
        for(int i=m;i>=l;i--)
        {
            sum += nums[i];
            left_sum = max(sum,left_sum);
        }
        int right_sum = INT_MIN;
        sum=0;
        for(int i=m+1;i<=r;i++)
        {
            sum += nums[i];
            right_sum = max(sum,right_sum);
        }
        return left_sum + right_sum;
    }
    int maxSum(int l,int r,vector<int>& nums)
    {
        if(l == r)
            return nums[l];
        int m = (l+r) >> 1;
        return max(max(maxSum(l,m,nums),maxSum(m+1,r,nums)),cross(l,m,r,nums));
    }
    int maxSubArray(vector<int>& nums) {
       return maxSum(0,nums.size()-1,nums);
    }
    
};

File: 30-day-leetcoding-challenge/Maximum Subarray.java
class Solution {
    
    private int maxCross(int[] nums,int l,int m,int r)
    {
        int sum = 0,leftSum = Integer.MIN_VALUE,rightSum = Integer.MIN_VALUE;
        for(int i=m;i>=l;i--)
        {
            sum += nums[i];
            if(sum > leftSum)
                leftSum = sum;
        }
        sum = 0;
        for(int i=m+1;i<=r;i++)
        {
            sum += nums[i];
            if(sum > rightSum)
                rightSum = sum;
        }
        
        return leftSum + rightSum;
    }
    
    private int maxSubArray(int[] nums,int l,int r)
    {
        if(l == r)
            return nums[l];
        int m = (l+r)/2;
        
        return Math.max(
            maxCross(nums,l,m,r),
            Math.max(
                maxSubArray(nums,l,m),
                maxSubArray(nums,m+1,r)
            )
        );
    }
    
    public int maxSubArray(int[] nums) {
           return maxSubArray(nums,0,nums.length-1);
    }
}
